napari/_qt/threading.py

```diff
@@ -0,0 +1,842 @@
+        # polling interval: ONLY relevant if the user paused a running worker
+        self._pause_interval = 0.01
+
+        if init_yield:
+            first_yield = next(self._gen)
+            if isinstance(first_yield, dict):
+                self.parse_first_yield(first_yield)
+
+    def parse_first_yield(self, first_yield: dict):
+        self.__dict__.update(**first_yield)
+
+    def work(self) -> None:
+        """Core event loop that calls the original function.
+
+        Enters a continual loop, yielding and returning from the original
+        function.  Checks for various events (quit, pause, resume, etc...).
+        (To clarify: we are creating a rudimentary event loop here because
+        there IS NO Qt event loop running in the other thread to hook into)
+        """
+        while True:
+            if self.abort_requested:
+                self.aborted.emit()
+                break
+            if self._paused:
+                if self._resume_requested:
+                    self._paused = False
+                    self._resume_requested = False
+                    self.resumed.emit()
+                else:
+                    time.sleep(self._pause_interval)
+                    continue
+            elif self._pause_requested:
+                self._paused = True
+                self._pause_requested = False
+                self.paused.emit()
+                continue
+            try:
+                self.pre_yield_hook()
+                self.yielded.emit(self._gen.send(self._next_value()))
+                self.post_yield_hook()
+            except StopIteration as exc:
+                return exc.value
+
+    def send(self, value: Any):
+        """Send a value into the function (if a generator was used)."""
+        self._incoming_value = value
+
+    def _next_value(self) -> Any:
+        out = None
+        if self._incoming_value is not None:
+            out = self._incoming_value
+            self._incoming_value = None
+        return out
+
+    @property
+    def is_paused(self) -> bool:
+        """Whether the worker is currently paused."""
+        return self._paused
+
+    def toggle_pause(self) -> None:
+        """Send a request to pause the worker if playing or resumed if paused.
+        """
+        if self.is_paused:
+            self._resume_requested = True
+        else:
+            self._pause_requested = True
+
+    def pause(self) -> None:
+        """Send a request to pause the worker.
+        """
+        if not self.is_paused:
+            self._pause_requested = True
+
+    def resume(self) -> None:
+        """Send a request to resume the worker.
+        """
+        if self.is_paused:
+            self._resume_requested = True
+
+    def pre_yield_hook(self):
+        """Hook for subclasses. Called just before yielding from generator"""
+        pass
+
+    def post_yield_hook(self):
+        """Hook for subclasses. Called just after yielding from generator"""
+        pass
```

Comment:
Are these used now that there is no progressworker? If not, I would also remove them from the PR.